src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoCommon.cs

```diff
@@ -69,19 +80,52 @@ internal override object GetValueAsObject(object obj, JsonSerializerOptions opti
                 return obj;
             }
 
-            Debug.Assert(Get != null);
+            Debug.Assert(HasGetter);
             return Get((TClass)obj);
         }
 
         internal override void SetValueAsObject(object obj, object value, JsonSerializerOptions options)
         {
-            Debug.Assert(Set != null);
+            Debug.Assert(HasSetter);
             TDeclaredProperty typedValue = (TDeclaredProperty)value;
 
             if (typedValue != null || !IgnoreNullPropertyValueOnWrite(options))
             {
-                Set((TClass)obj, (TDeclaredProperty)value);
+                Set(obj, typedValue);
             }
         }
+
+        internal TDeclaredProperty Get(object obj)
+        {
+            if (default(TClass) == null)
+            {
+                var get = (GetPropertyValue)_get;
+                return get((TClass)obj);
+            }
+            else
+            {
+                var get = (GetPropertyValueByRef)_get;
+                return get(ref Unsafe.As<ByReference>(obj).Value);
+            }
+        }
+
+        internal void Set(object obj, TDeclaredProperty value)
```

Comment:
Isn't the struct boxed on each setter? So the new property value won't be applied to `obj`?

Ideally this is `ref TClass obj` for structs so we don't box (and unbox) but that requires re-factoring of other calling code. If we change it to `ref object obj` and for structs below cast `obj` to `TClass` and then re-assign that back to obj that would work but slow (unbox + box)

Reply:
It's already boxed and stored on the heap. `object` means a reference to a `class` instance or a reference to a boxed value type located on the heap too. `ref object` is a pointer to a local which stores a pointer to the actual object.

Reply:
Yes the my sentence was because I wasn't sure the `ByReference` class was sufficient.